pdf-cdn.md

```diff
@@ -0,0 +1,222 @@
+
+Google serves `astro-ph/0606003v1.pdf`: 218ms wait for response, 51ms
+download. This is going directly without any redirects.
+
+These times omit DNS and SSL. These don't need to be top level
+concerns but we do expect users to come without keep alive from email
+clients. To arxiv.org I'm seeing these as around DNS 54ms, initial
+connection+ssl: 135ms.  To google I'm seeing 50ms, initial
+connection+ssl: 95ms.
+
+These times are not averages, they are for single requests.
+
+## Proposal that meets all goals (Rejected)
+It seems the goal of no-code conflicts with goals of no redundancy and
+current version. I'm not finding a way with the Google products to
+make this happen without having a service.
+
+The closest I can see is: We serve the PDFs from a Google Storage
+bucket, via a load balancer and CDN. The objects in the bucket are
+arranged so a request for the current version gets the PDF that is
+current. This object will need to be updated as the paper is revised.
+
+Consider a paper with 3 versions we'd have to have the GS bucket like this:
+
+    https//pdf.arxiv.org/pdfs/1111.22222.pdf   | gs://pdf-bucket/pdfs/1111.22222.pdf   |
+    https//pdf.arxiv.org/pdfs/1111.22222v1.pdf | gs://pdf-bucket/pdfs/1111.22222v1.pdf |
+    https//pdf.arxiv.org/pdfs/1111.22222v2.pdf | gs://pdf-bucket/pdfs/1111.22222v2.pdf |
+    https//pdf.arxiv.org/pdfs/1111.22222v3.pdf | gs://pdf-bucket/pdfs/1111.22222v3.pdf |
+
+This would meet all goals except redundant copies and not all of the
+long standing URL patterns.
+
+To meet 6 maintaining long standing URL patterns we'd need more
+redundancy. This is due to the lack of flexibility in URL mapping in
+the LB. 
+
+I don't see an API that says `put_content_at_CDN_key(key, content)`. We can think of a service as that API. The service gets a request from the CDN and then it can be programmed to put any content there.
+
+I suggest we reject this proposal. Reason #1 is the goal of no-code,
+GCP only-products forces limitations on what URL to file mappings are
+possible due to the limitations of the GCP products. Reason #2 is We
+would end up with a system where the state of the bucket was constantly
+changing to support the URL mapping. The storage and the keys used
+should be limited to the concerns of storage and not have to also
+carry the concerns of URL mapping. Reason #3 it would be easier to
+test an app returns the correct file for a URL than to test
+CDN/LB/GS/obj-naming-conventions.
+
+A small app the mediates the URL to object mapping is a large gain of
+flexibility for little additional complexity. The state of the objects
+in the GS becomes much simpler.
+
+## End Game Proposal 
+This is a proposal for what would be ideal without any legacy concerns other than URL patterns.
+
+1. PDFs in GS with version numbers
+1. arxiv.org presents URLs to PDFs like https://arxiv.org/pdf/1234.12345.pdf
+1. The `/pdf` path URL mapped at the LB a backend with a CDN `pdf-service` an app running on Cloud Run
```

Comment:
What if this wasn't forwarded, but used a redirect instead? 
We could make all pdf urls at arxiv.org be redirects.
And add two subdomains: pdf.{cornell,gcp}.arxiv.org.
If we never allow the subdomains to have a non-versioned pdf url, that might help with caching, etc.

Reply:
After reading the next section, "migration proposal", that sounds like the same thing...

Reply:
A redirect allows the client to hit the CDN node closest to them.  
By forward do you mean something like reverse proxy?

Always redirecting to a version could be useful. I'm concerned there is some use case like "Copy and paste from the URL bar of my browser and not have a versioned or strange URL to the PDF"

Reply:
Yes, I was thinking of "forwarded" as a gateway or reverse proxy.

That makes sense- so if the url is awkward looking, or we may have it split over multiple locations...
So I guess we also need version urls that redirect to the right place. Maybe someday the abs page can have a "share" widget to generate the shortest url.

Reply:
I just checked out of curiosity, there's no (arx).iv top-level domain.